receiver/awsxrayreceiver/internal/udppoller/poller.go

```diff
@@ -0,0 +1,217 @@
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the Licen
+
+package udppoller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"go.uber.org/zap"
+
+	recvErr "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/errors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/socketconn"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/util"
+)
+
+const (
+	// Transport is the network transport protocol used
+	// by the poller
+	Transport = "udp"
+
+	// size of the buffer used by each poller.
+	// https://github.com/aws/aws-xray-daemon/blob/master/pkg/cfg/cfg.go#L182
+	// https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L171
+	pollerBufferSizeKB = 64 * 1024
+)
+
+// Poller represents one or more goroutines that are
+// polling from a UDP socket
+type Poller interface {
+	SegmentsChan() <-chan RawSegment
+	Start(receiverLongTermCtx context.Context)
+	Close() error
+}
+
+// RawSegment represents a raw X-Ray segment document.
+type RawSegment struct {
+	// Payload is the raw bytes that represent one X-Ray segment.
+	Payload []byte
+	// SegmentCtx is the short-lived context created per raw segment received
+	SegmentCtx context.Context
+}
+
+// Config represents the configurations needed to
+// start the UDP poller
+type Config struct {
+	ReceiverInstanceName string
+	Transport            string
+	Endpoint             string
+	NumOfPollerToStart   int
+}
+
+type poller struct {
+	receiverInstanceName string
+	udpSock              socketconn.SocketConn
+	logger               *zap.Logger
+	wg                   sync.WaitGroup
+	receiverLongLivedCtx context.Context
+	maxPollerCount       int
+	// closing this channel will shutdown all goroutines
+	// within this poller
+	shutDown chan bool
+
+	// all segments read by the poller will be sent to this channel
+	segChan chan RawSegment
+}
+
+// New creates a new UDP poller
+func New(cfg *Config, logger *zap.Logger) (Poller, error) {
+	if cfg.Transport != Transport {
+		return nil, fmt.Errorf(
+			"X-Ray receiver only supports ingesting spans through UDP, provided: %s",
+			cfg.Transport,
+		)
+	}
+
+	addr, err := net.ResolveUDPAddr(Transport, cfg.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+	sock, err := net.ListenUDP(Transport, addr)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("Listening on endpoint for X-Ray segments",
+		zap.String(Transport, addr.String()))
+
+	return &poller{
+		receiverInstanceName: cfg.ReceiverInstanceName,
+		udpSock:              sock,
+		logger:               logger,
+		maxPollerCount:       cfg.NumOfPollerToStart,
+		shutDown:             make(chan bool),
+		segChan:              make(chan RawSegment, 30),
```

Comment:
The size of this channel may be something that you want to make configurable.

Reply:
Good suggestion! I normally would. However, this PR is essentially a port of the functionalities of the existing X-Ray Daemon, so I'm trying to minimize new configurations that customers may not be familiar with. We can always add more configurations in the future.

I will pull this out as a constant at the top so it's more explicit that we are hard coding something here.